### PR TITLE
[TGA] Beam EOS score fix

### PR DESCRIPTION
**Pull request by uralik**
Fix for https://github.com/facebookresearch/ParlAI/issues/2058

Steps to reproduce and show how it differs with the fix are shown in the issue linked above.

The first commit adds EOS penalty in the advance func.

The second commit changes the scores returned by the nucleus and topk. Since after the first EOS sequence will have a neginf score (since we panelize it in the base class), we change it to always output score of the currently selected token instead. Since there is no notion of a hypothesis or sequence in topk/nucleus, this should be fine. 

In case if one needs a seq-level score, it can be easily obtained by summing over needed scores from `self.all_scores` (no backtracking is needed as in beam).

### CODE REVIEW

File: parlai/core/torch_generator_agent.py [hunk 1]
```diff
@@ -1043,8 +1049,7 @@ def select_paths(self, logprobs, prior_scores):
         hyp_ids = torch.arange(logprobs.size(0)).to(logprobs.device)
         tok_ids = indices[hyp_ids, choices]
         scores = values[hyp_ids, choices]
-        best_scores = prior_scores.expand_as(scores) + scores
-        return (hyp_ids, tok_ids, best_scores)
+        return (hyp_ids, tok_ids, scores)
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: Why are you changing this? The contract of tree search is the logprob of the full utterance is contained in the third return value.

File: parlai/core/torch_generator_agent.py [hunk 2]
```diff
@@ -853,8 +852,15 @@ def advance(self, logprobs):
...
 
+        # penalize hypotheses ending in EOS on the prior scores (self.scores) level
+        # this is only related to search which uses prior scores (self.scores) (e.g. beam)
+        for hyp_id, token in enumerate(self.outputs[-1]):
+            if token == self.eos:
+                self.scores[hyp_id] = neginf(self.scores.dtype)
+
         hyp_ids, tok_ids, self.scores = self.select_paths(logprobs, self.scores)
-        self.all_scores.append(self.scores)
+        # use clone() here to ensure that self.all_scores will not be changed later due to any penalties to self.scores
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: wrap on 2 lines please

File: parlai/core/torch_generator_agent.py [hunk 3]
```diff
@@ -870,7 +876,7 @@ def advance(self, logprobs):
                 eostail = _HypothesisTail(
                     timestep=len(self.outputs) - 1,
                     hypid=hypid,
-                    score=self.scores[hypid],
+                    score=self.all_scores[-1][hypid],
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: does this mean we're excluding the probability of generating the EOS token?

File: parlai/core/torch_generator_agent.py [hunk 4]
```diff
@@ -8,8 +8,7 @@
 """
 Generic PyTorch-based Generator agent.
 
-Implements quite a bit of boilerplate, including forced-decoding loss and beam
-search.
+Implements quite a bit of boilerplate, including forced-decoding loss and a tree search.
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: does this still fit on 88 characters?

File: parlai/core/torch_generator_agent.py [hunk 5]
```diff
@@ -853,8 +852,15 @@ def advance(self, logprobs):
         if self.scores is None:
             self.scores = torch.zeros(1).type_as(logprobs).to(logprobs.device)
 
+        # penalize hypotheses ending in EOS on the prior scores (self.scores) level
+        # this is only related to search which uses prior scores (self.scores) (e.g. beam)
+        for hyp_id, token in enumerate(self.outputs[-1]):
+            if token == self.eos:
+                self.scores[hyp_id] = neginf(self.scores.dtype)
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: correct me if i'm wrong: if you modified `logprobs[hyp_id]` here such that EVERYTHING was neginf except padding, then the assert I propose would be okay. (Nucleus and topk search would start only outputting padding)

File: parlai/core/torch_generator_agent.py (hunk 3 above)
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: we want to put the value from all_scores here (copied there at line 863), because self.scores is an active buffer which we will penalize on the next time step. Then here it will be a -inf score in the end (this is bad).

File: parlai/core/torch_generator_agent.py (hunk 1 above)
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: I changed this to avoid having -inf scores here after the first EOS was sampled. But actually thats fine if all finalized hypos after the first one will have -inf score, I will change this back. Then only the first sampled EOS will make a non -inf finalized hypo per each beam id

File: parlai/core/torch_generator_agent.py (hunk 5 above)
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: discussed offline another way

File: parlai/core/torch_generator_agent.py (hunk 4 above)
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: yes

File: parlai/core/torch_generator_agent.py (hunk 2 above)
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: ok

**Comment by uralik** [comment]
Several changes in the latest commit:

1.  Only EOS pointers with legit (non-inf) scores are added to `.finished`, this way to make sure that in case of stochastic decoding all finished predictions have only a single EOS.
2. To check point 1 above there is a loop with assert over finalized candidates
3. Since this loop in point 2 are done, there is no need to call `.get_top_hyp`, it is a waste of time, we change this too.
4. Now important: with such assert, default value of `beam_min_n_best` (3) is not good anymore. For example, in such case any decoding with beam size < 3 will crash with this assert. To address this, we add:
a. default argparse value now is 0. If it was not given form saved opt, then we substitute beam_size value there in the code (safe option if a user has no idea about it).
b. if we load the model and value there will be 3 (old default), then we put a warning saying that in some cases this may lead to a crash. In the comment we also point out that user may override this with a CLI argument.

Perhaps there are better ways of handling this `beam_min_n_best`, lets discuss if you have smth in mind.

File: parlai/core/torch_generator_agent.py [hunk 6]
```diff
@@ -305,9 +297,8 @@ def __init__(self, opt, shared=None):
         init_model, is_finetune = self._get_init_model(opt, shared)
         super().__init__(opt, shared)
 
-        self.beam_size = opt.get('beam_size', 1)
-        self.beam_min_n_best = opt.get('beam_min_n_best', 3)
-        self.beam_min_length = opt.get('beam_min_length', 3)
+        self.beam_size = opt.get('beam_size')
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: These changes should be undone. We don’t want none values

File: parlai/core/torch_generator_agent.py [hunk 7]
```diff
@@ -264,17 +263,10 @@ def add_cmdline_args(cls, argparser):
...
-            type=int,
-            default=3,
-            help='Minimum number of nbest candidates to achieve '
-            'during the beam search',
-        )
         agent.add_argument(
             '--beam-min-length',
             type=int,
-            default=1,
+            default=3,
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: Undo this

File: parlai/core/torch_generator_agent.py [hunk 8]
```diff
@@ -728,8 +715,21 @@ def _generate(self, batch, beam_size, max_ts):
             ).unsqueeze(-1)
             decoder_input = torch.cat([decoder_input, selection], dim=-1)
 
-        # get the top prediction for each
-        beam_preds_scores = [b.get_top_hyp() for b in beams]
+        # check that there is at least one finished candidate
+        # and assert that each of them contains only one EOS
+        n_best_beam_preds_scores = [b._get_rescored_finished() for b in beams]
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: So we don’t want to call methods with a leading _, because they’re supposed to be protected. So can we upgrade the behavior of get_top_hyp instead of having this logic here?

File: parlai/core/torch_generator_agent.py [hunk 9]
```diff
@@ -942,12 +947,13 @@ def _get_rescored_finished(self, n_best=None):
...
         # if we never actually finished, force one
         if not self.finished:
+            self.outputs[-1][0] = self.eos
             self.finished.append(
                 _HypothesisTail(
                     timestep=len(self.outputs) - 1,
                     hypid=0,
                     score=self.all_scores[-1][0],
-                    tokenid=self.eos,
+                    tokenid=self.outputs[-1][0],
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: Do we need these changes?

Comment (uralik) [reply parlai/core/torch_generator_agent.py]: basically we can keep self.eos, but using outputs (previously updated) makes a clear reason why we have updated it, so I would keep it like this

File: parlai/core/torch_generator_agent.py (hunk 8 above)
Comment (uralik) [reply parlai/core/torch_generator_agent.py]: yeah, we get rid of get_top_hyp since we actually do not use it at all , since we do this EOS assert check for each candidate. Now get_rescored_finished will be not protected.
